test/integration/load/load_test.go

```diff
@@ -0,0 +1,161 @@
+	restartCase       = flag.Bool("restart-case", false, "In restart case, skip if we don't find state file")
+	namespace         = "load-test"
+)
+
+/*
+In order to run the scale tests, you need a k8s cluster and its kubeconfig.
+If no kubeconfig is passed, the test will attempt to find one in the default location for kubectl config.
+Run the tests as follows:
+
+go test -timeout 30m -tags load -run ^TestLoad$ -tags=load
+
+The Load test scale the pods up/down on the cluster and validates the pods have IP. By default it runs the
+cycle for 2 iterations.
+
+To validate the state file, set the flag -validate-statefile to true. By default it is set to false.
+todo: consider adding the following scenarios
+- [x] All pods should be assigned an IP.
+- [x] Test the CNS state file.
+- [x] Test the CNS Local cache.
+- [x] Test the Cilium state file.
+- [x] Test the Node restart.
+- [ ] Test based on operating system.
+- [ ] Test the HNS state file.
+- [ ] Parameterize the os, cni and number of iterations.
+- [ ] Add deployment yaml for windows.
+*/
+func TestLoad(t *testing.T) {
+	clientset, err := k8sutils.MustGetClientset()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Minute)
+	defer cancel()
+
+	err = k8sutils.MustCreateNamespace(ctx, clientset, namespace)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	deployment, err := k8sutils.MustParseDeployment(noopdeployment)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	deploymentsClient := clientset.AppsV1().Deployments(namespace)
+	err = k8sutils.MustCreateDeployment(ctx, deploymentsClient, deployment)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Log("Checking pods are running")
+	err = k8sutils.WaitForPodsRunning(ctx, clientset, namespace, podLabelSelector)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Log("Repeating the scale up/down cycle")
+	for i := 0; i < *iterations; i++ {
+		t.Log("Iteration ", i)
+		t.Log("Scale down deployment")
+		err = k8sutils.MustScaleDeployment(ctx, deploymentsClient, deployment, clientset, namespace, podLabelSelector, *scaleDownReplicas, *skipWait)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Log("Scale up deployment")
+		err = k8sutils.MustScaleDeployment(ctx, deploymentsClient, deployment, clientset, namespace, podLabelSelector, *scaleUpReplicas, *skipWait)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	t.Log("Checking pods are running and IP assigned")
+	err = k8sutils.WaitForPodsRunning(ctx, clientset, "", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if *validateStateFile {
+		t.Run("Validate state file", TestValidateState)
+	}
+}
+
+// TestValidateState validates the state file based on the os and cni type.
+func TestValidateState(t *testing.T) {
+	clientset, err := k8sutils.MustGetClientset()
+	if err != nil {
+		t.Fatal(err)
+	}
+	config := k8sutils.MustGetRestConfig(t)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Minute)
```

Comment:
Is there any reasoning for the 30 minute timeout in both TestValidateState and TestLoad? I think we should consider parameterizing at least TestLoad's timeout for future tests and keep the default to 30 minutes with a flag.

Reply:
I tried testing without the timeout context for LoadTest to get an understanding of how long does it takes to scale the pods to max - the scaling was done in around 25 mins for around 240 pods per node( max we can have is 250). That is why 30 mins.
By future tests, I believe you are pointing towards scaling on different cni/os versions right ? I think since all the scaling of pods will be happening parallelly on different cluster, adding another parameter for increasing context would be too much right now.(Already have a bunch of them)
Are any there any cases you want to tackle by making it a parameter ?

Reply:
For future cases I imagine the pod threshold to be higher, new features that slow down the throughput i.e. cilium, and possible timing differences between windows and linux networking. 

Considering it is able to complete the test 4 times in 25 minutes this should be enough time for future tests.